### DISCUSSION TITLE
Spherical harmonics coil design example

### DISCUSSION THREAD

**Discussion by GBZH** [bfieldtools/bfieldtools#62] [Q&A]
Hello,

Looking at the spherical harmonics coil design example, I have 2 questions:

1) I am wondering about the following lines:

```
target_alms = np.zeros((coil.opts["N_sph"] * (coil.opts["N_sph"] + 2),))
target_blms = np.zeros((coil.opts["N_sph"] * (coil.opts["N_sph"] + 2),))
target_blms[4] += 1
```

By default N_sph = 5, therefore it means that both the arrays target_alms and target_blms have 35 elements. If one assumes that A_lm (resp. B_lm) are the real harmonics coefficients associated with cos(m phi) (resp. sin(m phi)) in the spherical harmonics series expansion, I do not understand why they are so long and how they are ordered versus degree l and order m? Id est, what is the correspondence between target_alms[0] , ..., target_alms[i], ..., target_alms[34] (i ∈ [0, 34]) and A_lm for a given l and a given m? Same question for target_blms[i] and B_lm? 
Should we ignore one of these arrays as each of them stores harmonics coefficients in the following order:
A_1_0,  A_1_1, B_1_1, 
A_2_0, A_2_1, B_2_1, A_2_2, B_2_2,
...
A_5_0, A_5_1, B_5_1, ..., A_5_5, B_5_5,
ignoring terms A_0_0 and all B_l_0 w/ l ∈ [0, 5]?

2) Then, in the optimisation process below, why the target specification only takes into account
target_blms and not target_alms ?

```
coil.set_sph_options(sph_radius=0.5)

target_spec = {
    "coupling": coil.sph_couplings[1],
    "abs_error": 0.001,
    "target": target_blms,
}
```
Thanks a lot for any help,

Gilles

**Comment by phuonganhlovely** [comment]
what application do you want to design the coil for?
I am looking to design a coil for electric machine ( motor) base on spherical harmonic. do you have any ideas for this topic?

**Reply by GBZH** [reply]
Concerning the way harmonics are stored in the code, I had to dig into the code and it is different than what I thought: it is in increasing order of l and for each l in increasing order of m.

I am designing coils for some fundamental physics experiment. Coil giviving a particular harmonic field at a given place and low field at other location. I am currently working on it, but I am not yet satisfied by my first results.

**Comment by jiivana** [comment]
Hi GBZH,

Thanks for the good question. The ordering of the sph functions is indeed a little bit vague and should be probably explained better. But you nevertheless found the correct answer.

alms and blms correspond to inner and outer expansion components of the spherical harmonics expansion of the magnetic field. alm sphs diverge at origin and blm sphs converge at origin. When your volume of interest is around origin and coils around the volume of interest, blm sph expansion should be used as it describes the magnetic field from outside sources.
